mipengine/controller/algorithm_executor.py

```diff
@@ -0,0 +1,583 @@
+    def get_remote_tables(self) -> List["TableInfo"]:
+        return self.node_tasks_handler.get_remote_tables(context_id=self.context_id)
+
+    def create_remote_table(
+        self, table_info: TableInfo, native_node: _Node
+    ) -> _TableName:
+
+        monetdb_socket_addr = native_node.node_address
+        return self.node_tasks_handler.create_remote_table(
+            table_info=table_info, original_db_url=monetdb_socket_addr
+        )
+
+    # UDFs functionality
+    def queue_run_udf(
+        self, command_id: str, func_name: str, positional_args, keyword_args
+    ) -> "AsyncResult":
+        return self.node_tasks_handler.queue_run_udf(
+            context_id=self.context_id,
+            command_id=command_id,
+            func_name=func_name,
+            positional_args=positional_args,
+            keyword_args=keyword_args,
+        )
+
+    def get_udfs(self, algorithm_name) -> List[str]:
+        return self.node_tasks_handler.get_udfs(algorithm_name)
+
+    # return the generated monetdb pythonudf
+    def get_run_udf_query(
+        self, command_id: str, func_name: str, positional_args: List[_NodeTable]
+    ) -> Tuple[str, str]:
+        return self.node_tasks_handler.get_run_udf_query(
+            context_id=self.context_id,
+            command_id=command_id,
+            func_name=func_name,
+            positional_args=positional_args,
+        )
+
+    # CLEANUP functionality
+    def clean_up(self):
+        self.node_tasks_handler.clean_up(context_id=self.context_id)
+
+
+class _AlgorithmExecutionInterfaceDTO(BaseModel):
+    global_node: _Node
+    local_nodes: List[_Node]
+    algorithm_name: str
+    algorithm_parameters: Dict[str, List[str]]
+    x_variables: List[str]
+    y_variables: List[str]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class _AlgorithmExecutionInterface:
+    def __init__(self, algo_execution_interface_dto: _AlgorithmExecutionInterfaceDTO):
+        self._global_node = algo_execution_interface_dto.global_node
+        self._local_nodes = algo_execution_interface_dto.local_nodes
+        self._algorithm_name = algo_execution_interface_dto.algorithm_name
+        self._algorithm_parameters = algo_execution_interface_dto.algorithm_parameters
+        self._x_variables = algo_execution_interface_dto.x_variables
+        self._y_variables = algo_execution_interface_dto.y_variables
+
+        # TODO: validate all local nodes have created the base_view_table??
+        self._initial_view_tables = {}
+        tmp_variable_node_table = {}
+
+        # TODO: clean up this mindfuck??
+        for node in self._local_nodes:
+            for (variable_name, table_name) in node.initial_view_tables.items():
+                if variable_name in tmp_variable_node_table:
+                    tmp_variable_node_table[variable_name].update({node: table_name})
+                else:
+                    tmp_variable_node_table[variable_name] = {node: table_name}
```

Comment:
what you propose is more readable but I think it has to be even simpler, but it will need some refactoring on other parts of the code. This was not introduced in this pr/commit, it has been there from the beginning. So I added a link to your comment and I ll refactoring later...